lib/puppet/pops/types/type_calculator.rb

```diff
@@ -452,9 +452,18 @@ def instance_of_PTupleType(t, o)
 
   def instance_of_PStructType(t, o)
     return false unless o.is_a?(Hash)
-    h = t.hashed_elements
-    # all keys must be present and have a value (even if nil/undef)
-    (o.keys - h.keys).empty? && h.all? { |k,v| instance_of(v, o[k]) }
```

Comment:
To be honest, it didn't start as an optimization. I didn't realize that nil was assignable to Optional, hence the change. I can move it into a separate commit, but doesn't that also imply creating a separate issue and a separate PR? If so, then let's just forget about it. It's not worth all that.